### PR TITLE
feat(app-framework): support undo for intents

### LINKED ISSUE #5041: [composer] undo plugin
- [ ] Undo plugin
- [ ] Toast appears for 5s to enable undo

### PR DESCRIPTION
**Pull request by wittjosiah**
### Motivation / Background

This PR adds history to the intent plugin in order to be able to support undo.

Resolves #5041

### Details

- update intent types to handle the possibility of inverse intents
- add intent for when undo is available
- layout plugin resolves available undos to a toast message
- test undo by making thread deletion undoable

### Additional information

- At present undo only supports one action but history is total, so in the future you should be able to browse the history as well as undo several actions.
- Initially implemented via a toast, but expectation is that this will have custom UI designed for it in the future as well.
- [ ] remove old undo plugin

Screenshot 2024-02-05 at 14 09 11

### CODE REVIEW

File: packages/apps/plugins/plugin-thread/src/ThreadPlugin.tsx [hunk 1]
```diff
@@ -221,6 +223,25 @@ export const ThreadPlugin = (): PluginDefinition<ThreadPluginProvides> => {
...
+              const { document: doc, thread, cursor } = intent.data ?? {};
+              if (!isDocument(doc) || !isThread(thread)) {
+                return;
+              }
+
+              if (intent.inverse && typeof cursor === 'string') {
+                doc.comments.push({ thread, cursor });
+                return { data: true };
+              } else {
+                const index = doc.comments.findIndex((comment) => comment.thread?.id === thread.id);
```
Comment (richburdon) [inline packages/apps/plugins/plugin-thread/src/ThreadPlugin.tsx:237]: @dmaretskyi should we actively delete the object and not just splice it (and restore above)?

File: packages/sdk/app-framework/src/plugins/IntentPlugin/intent.ts [hunk 2]
```diff
@@ -30,6 +30,11 @@ export type Intent = {
    */
   action: string;
 
+  /**
+   * Whether or not the intent is being undone.
+   */
+  inverse?: boolean;
```
Comment (richburdon) [inline packages/sdk/app-framework/src/plugins/IntentPlugin/intent.ts:36]: i think i prefer `undo`

File: packages/sdk/app-framework/src/plugins/IntentPlugin/provides.ts [hunk 3]
```diff
@@ -21,3 +21,15 @@ export const parseIntentPlugin = (plugin: Plugin) =>
...
+//
+// Intents
+//
+
+const INTENT_ACTION = 'dxos.org/plugin/intent';
+export enum IntentAction {
+  /**
+   * Fired after an intent is dispatched if the intent is undoable.
+   */
+  UNDO_AVAILABLE = `${INTENT_ACTION}/undo-available`,
```
Comment (richburdon) [inline packages/sdk/app-framework/src/plugins/IntentPlugin/provides.ts:34]: I think `SHOW_UNDO` is more consistent. unless there's an even broader intent such as HISTORY_UPDATED where the layout would get the current state of the history (incl. undo being available).

File: packages/apps/plugins/plugin-thread/src/ThreadPlugin.tsx [hunk 4]
```diff
@@ -221,6 +223,25 @@ export const ThreadPlugin = (): PluginDefinition<ThreadPluginProvides> => {
...
               return { data: true };
             }
+
+            case ThreadAction.DELETE: {
+              const { document: doc, thread, cursor } = intent.data ?? {};
+              if (!isDocument(doc) || !isThread(thread)) {
+                return;
+              }
+
+              if (intent.inverse && typeof cursor === 'string') {
```
Comment (richburdon) [inline packages/apps/plugins/plugin-thread/src/ThreadPlugin.tsx:233]: it feels a bit error prone to `else` here if EITHER of these are false. feels like instead cursor === string should be an invariant inside the inverse (undo) case?

**Review by richburdon — COMMENTED** [review]
nicely done

File: packages/apps/plugins/plugin-thread/src/ThreadPlugin.tsx (hunk 1 above)
Comment (wittjosiah) [reply packages/apps/plugins/plugin-thread/src/ThreadPlugin.tsx:237]: @richburdon note that comments are not standalone objects at the moment but apart of the the `Document` schema. Only `Thread`s are separate objects.

File: packages/apps/plugins/plugin-layout/src/translations.ts [hunk 5]
```diff
@@ -20,6 +20,10 @@ export default [
         'toggle fullscreen label': 'Toggle fullscreen',
         'settings enable complementary sidebar label': 'Show context sidebar.',
         'settings show footer label': 'Show footer.',
+        'undo available label': 'Click to undo previous action.',
```
Comment (thure) [inline packages/apps/plugins/plugin-layout/src/translations.ts:23]: I don’t suppose we might be able to provide the user with more information than this? I think ideally it’d say something like:

> You deleted a thread.
> Undo | Dismiss

Comment (thure) [reply packages/apps/plugins/plugin-layout/src/translations.ts:23]: Can sort it out in a follow-up if we get feedback indicating folks would want this, I know it’s potentially a lot of added complexity.

Comment (wittjosiah) [reply packages/apps/plugins/plugin-layout/src/translations.ts:23]: Good point, let me see if I can find a way that we could have a custom label here specified by the action.

**Review by thure — APPROVED** [review]
Meant to approve, especially so if we can tell the user what triggered the toast in the toast, but that can come later.

**Pull request merged**